### PR TITLE
Implement Khatri-Rao operator

### PR DESCRIPTION
**Pull request by cswiercz**
Create an operator for the Khatri-Rao function implemented in #6567. The operator accepts a variable number of input matrix arguments as well as a flag indicating whether the Khatri-Rao product should be computed row-wise or column-wise. This option is provided because the row-wise operation is more memory efficient.

### A Note to Reviewers

This is my first time putting MXNet code up for review so please tear this code apart. I need to learn.

Additionally, I've knowingly omitted the Backwards/Gradient calculation.

### CODE REVIEW

**Comment by cswiercz** [comment]
Pinging @jli05 and @JeanKossaifi

**Comment by JeanKossaifi** [comment]
@cswiercz that's a great addition and will be very useful! Was meaning to do it but didn't have time..

I am not sure about the row_wise flag. The khatri-rao is a column-wise kronecker product, and the row-wise would also refer to the kronecker product, not khatri-rao.. Would it be clearer to have khatri_rao and maybe row_wise_kronecker (long name but unambiguous?)

**Comment by cswiercz** [comment]
> I am not sure about the row_wise flag. The khatri-rao is a column-wise kronecker product, and the row-wise would also refer to the kronecker product, not khatri-rao.. Would it be clearer to have khatri_rao and maybe row_wise_kronecker (long name but unambiguous?)

Yeah, I wasn't so sure about the flag. Having a separate operator sounds like a better idea. For consistency maybe:

* `khatri_rao()`
* `column_wise_kronecker()` *(alias to `khatri_rao`)*
* `row_wise_kronecker()`

Thoughts from anyone who is interested?

**Comment by jli05** [comment]
I'm for exposing only column-wise Khatri-Rao for the Python Op, but only reserving the row-wise one for framework developers (C++).

**Comment by cswiercz** [comment]
> I'm for exposing only column-wise Khatri-Rao for the Python Op, but only reserving the row-wise one for framework developers (C++).

Fair enough. It would certainly simplify the code considerably. I made the appropriate changes.

**Comment by cswiercz** [comment]
@mli @piiswrong - Two MXNet contributors who know this function have approved this PR. Is there anything I'm missing?

**Comment by cswiercz** [comment]
@mli @piiswrong - If it makes any difference, this code is going into `src/operator/contrib`.

File: src/operator/c_lapack_api.h [hunk 1]
```diff
@@ -111,7 +111,11 @@ inline char loup(char uplo, bool invert) { return invert ? (uplo == 'U' ? 'L' :
  * \param lda leading dimension of a
  */
 template <typename xpu, typename DType>
-inline void flip(int m, int n, DType *b, int ldb, DType *a, int lda);
```
Comment (piiswrong) [inline src/operator/c_lapack_api.h]: What was the previous definition for flip?

Comment (cswiercz) [reply src/operator/c_lapack_api.h]: Only the `DType=float` and `DType=double` versions were implemented in the subsequent lines. I the function prototype in this line with a generalized version templated on the `DType`.

This change was necessary due to the use of `MSHADOW_TYPE_SWITCH`, I believe. Without this templating I was observing compile errors when testing with non `float` and `double` types.

I think the `float` and `double`-specific versions can be removed, actually...

**Comment by cswiercz** [comment]
> Sorry for missing this. Please ping me with email if PR gets ignored for too long next time.

No worries. I know you're busy with more pressing, non-contrib things.

File: src/operator/c_lapack_api.h (hunk 1 above)
Comment (piiswrong) [reply src/operator/c_lapack_api.h]: Yes they don't make much sense.

Please remove them

Comment (cswiercz) [reply src/operator/c_lapack_api.h]: Done. See 66afdc3 below.

**Comment by piiswrong** [comment]
@cswiercz Could you rebase to current master?

**Comment by cswiercz** [comment]
@piiswrong Rebased. All tests pass.

**Pull request merged**